selectors/example-selectors.md

```diff
@@ -0,0 +1,218 @@
+Example Selectors
+=================
+
+In this document, we will show some examples of selectors.
+
+#### A quick note about formatting
+
+We will typically show some sample data as JSON encoded documents,
+and show the Selectors as YAML documents.
+**Remember,** any IPLD document can be encoded in any multicodec;
+these examples are JSON for convenience and by convention, but all
+the examples would work the same even if they were in other formats.
+Similarly, **Selectors themselves are just IPLD documents**, so while
+we've written yaml forms here, they could just as well be e.g. JSON or CBOR.
+When we write yaml, we'll also use quoted strings for user-supplied strings,
+and unquoted strings as map keys when it's for a field specified by the schema.
+
+For human convenience, we will also pretend there are a few modifications
+to the Selector schema:
+
+- we will disregard all 'rename' shortenings specified by the schema;
+- and we will use type names for union discriminators,
+  instead of the shorter keys specified in the schema for those unions.
+
+Therefore, note that the *real* serialized selectors will be significantly terser!
+
+
+Examples
+--------
+
+### Deeply nested path
+
+In some DAG you want to get one specific value you know the path of. Let's say you want to get the birth year of a specific character of a specific show.
+
+Example data (as JSON):
+
+```json
+{
+  "show": "start-trek-voyager",
+  "characters": {
+    "kathryn-janeway": {
+      "birthday": {
+        "day": 20,
+        "month": 5,
+        "year": 2328
+      },
+      "rank": "vice admiral"
+    }
+  }
+}
+```
+
+A Selector to extract the "year" data could look like this:
+
+```yaml
+Selector:
+  ExploreFields:
+    fields:
+      "characters":
+        ExploreFields:
+          fields:
+            "kathryn-janeway":
+              ExploreFields:
+                fields:
+                  "birthday":
+                    ExploreFields:
+                      fields:
+                        "year":
+                          Matcher:
+                            {}
```

Comment:
I guess this is correct, I'm just not sure that YAML is expressive enough to show what's going on here, although that could be that my eyes aren't as comfortable with YAML as some other formats.

Reply:
It ain't pretty, but yeah, it's correct.  It's just a very right-leaning tree.  Hard to make those pretty.

(Fwiw, one of the oldest iterations of this whole Selectors thing actually used lists because we were so eager to avoid the right-leaning tree it seemed like a viable option.  It was not: more complex queries that we instantly had users toss at us were inexpressible in that notation.)

Probably the prettiest format for something like this would take inspiration from XML: the keyed unions here (`ExploreFields`, `Matcher` in this example) are an awfully lot like the tag in XML, and the map inside those is an awfully lot like attributes.  (Or, are the inner maps like nested keys?  Ahha.  This is why we don't use XML... ;)  Fun to think about other syntaxes that would have better visual presentation though!)

Reply:
In my editor (vim) it looks much nicer, as the things in quotes are highlighted differently from the ones without quotes.
![image](https://user-images.githubusercontent.com/42839/66041503-47eaee80-e51a-11e9-89b4-b00cbc68311c.png)